core/arch/arm/mm/mobj.c

```diff
@@ -0,0 +1,284 @@
+	struct mobj mobj;
+	enum buf_is_attr battr;
+	uint32_t cattr; /* Defined by TEE_MATTR_CACHE_* in tee_mmu_types.h */
+	vaddr_t va;
+	paddr_t pa;
+};
+
+static struct mobj_phys *to_mobj_phys(struct mobj *mobj);
+
+static void *mobj_phys_get_va(struct mobj *mobj, size_t offset)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+
+	if (!moph->va)
+		return NULL;
+
+	return (void *)(moph->va + offset);
+}
+
+static TEE_Result mobj_phys_get_pa(struct mobj *mobj, size_t offs,
+				   size_t granule, paddr_t *pa)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+	paddr_t p;
+
+	if (!pa)
+		return TEE_ERROR_GENERIC;
+
+	p = moph->pa + offs;
+
+	if (granule) {
+		if (granule != SMALL_PAGE_SIZE &&
+		    granule != CORE_MMU_PGDIR_SIZE)
+			return TEE_ERROR_GENERIC;
+		p &= ~(granule - 1);
+	}
+
+	*pa = p;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mobj_phys_get_cattr(struct mobj *mobj, uint32_t *cattr)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+
+	if (!cattr)
+		return TEE_ERROR_GENERIC;
+
+	*cattr = moph->cattr;
+	return TEE_SUCCESS;
+}
+
+static bool mobj_phys_matches(struct mobj *mobj, enum buf_is_attr attr)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+	enum buf_is_attr a;
+
+	a = moph->battr;
+
+	switch (attr) {
+	case CORE_MEM_SEC:
+		return a == CORE_MEM_SEC || a == CORE_MEM_TEE_RAM ||
+		       a == CORE_MEM_TA_RAM;
+	case CORE_MEM_NON_SEC:
+		return a == CORE_MEM_NSEC_SHM;
+	case CORE_MEM_TEE_RAM:
+	case CORE_MEM_TA_RAM:
+	case CORE_MEM_NSEC_SHM:
+		return attr == a;
+	case CORE_MEM_MULTPURPOSE:
+		return a == CORE_MEM_TA_RAM || CORE_MEM_NSEC_SHM;
+	default:
+		return false;
+	}
+}
+
+static void mobj_phys_free(struct mobj *mobj)
+{
+	struct mobj_phys *moph = to_mobj_phys(mobj);
+
+	free(moph);
+}
+
+static const struct mobj_ops mobj_phys_ops __rodata_unpaged = {
+	.get_va = mobj_phys_get_va,
+	.get_pa = mobj_phys_get_pa,
+	.get_cattr = mobj_phys_get_cattr,
+	.matches = mobj_phys_matches,
+	.free = mobj_phys_free,
+};
+
+static struct mobj_phys *to_mobj_phys(struct mobj *mobj)
+{
+	assert(mobj->ops == &mobj_phys_ops);
+	return container_of(mobj, struct mobj_phys, mobj);
+}
+
+struct mobj *mobj_phys_alloc(paddr_t pa, size_t size, uint32_t cattr,
+			     enum buf_is_attr battr)
+{
+	struct mobj_phys *moph;
+	void *va;
+
+	va = phys_to_virt(pa, battr);
+	if (!va)
+		return NULL;
+
+	moph = calloc(1, sizeof(*moph));
+	if (!moph)
+		return NULL;
+
+	moph->battr = battr;
+	moph->cattr = cattr;
+	moph->mobj.size = size;
+	moph->mobj.ops = &mobj_phys_ops;
+	moph->pa = pa;
+	moph->va = (vaddr_t)va;
+
+	return &moph->mobj;
+}
+
+/*
+ * mobj_virt implementation
+ */
+
+static void mobj_virt_assert_type(struct mobj *mobj);
+
+static void *mobj_virt_get_va(struct mobj *mobj, size_t offset)
+{
+	mobj_virt_assert_type(mobj);
+
+	return (void *)(vaddr_t)offset;
+}
+
+static void mobj_virt_free(struct mobj *mobj)
+{
+	mobj_virt_assert_type(mobj);
+	free(mobj);
+}
+
+static const struct mobj_ops mobj_virt_ops __rodata_unpaged = {
```

Comment:
i don't catch this. no mobj->ops  is set to `mobj_virt_ops` by `mobj_virt_xxx` expect so on handled object.